manager/allocator/network/port/port.go

```diff
@@ -0,0 +1,451 @@
+		DynamicPortStart,
+		DynamicPortEnd,
+		e.protocol,
+	)
+}
+
+// IsErrPortSpaceExhausted returns true if the error is of type
+// ErrPortSpaceExhausted
+func IsErrPortSpaceExhausted(e error) bool {
+	_, ok := e.(ErrPortSpaceExhausted)
+	return ok
+}
+
+// ErrInvalidEndpoint is returned if the endpoint passed to Allocate or
+// Deallocate is somehow not valid. The error message will contain the problem.
+// It is a catchall for any errors not otherwise covered.
+type ErrInvalidEndpoint struct {
+	problem string
+}
+
+// Error returns a formatted message explaining what the problem is
+func (e ErrInvalidEndpoint) Error() string {
+	return e.problem
+}
+
+// IsErrInvalidEndpoint returns true if the provided error is of type
+// ErrInvalidEndpoint
+func IsErrInvalidEndpoint(e error) bool {
+	_, ok := e.(ErrInvalidEndpoint)
+	return ok
+}
+
+// Allocator is an allocator component that manages the state of which
+// ports (and which protocols) are in use in the the cluster.
+//
+// The Allocator totally owns the Endpoint.Ports field. It provides a slice
+// of PortConfigs which should be written directly without modification to the
+// the Endpoint.Ports field. Nothing should be modified about Endpoint.Ports
+// outside of the port allocator. If this rule is followed, the the port
+// allocator doesn't have to check for cases like two services having the same
+// port allocated; we created all of the fields, and we can know they are
+// consistent.
+type Allocator struct {
+	// ports maps the ports in use. essentially, using map as a set
+	ports map[port]struct{}
+}
+
+// port is the minimal representation for the Allocator of a single port.
+// being composed of 2 numeric types, this type is comparable and can be used
+// as a map key
+type port struct {
+	// protocol represented by this port space
+	protocol api.PortConfig_Protocol
+	port     uint32
+}
+
+// Proposal is the return value of Allocate and DeallocateEndpoint,
+// which contains the data needed to affirmatively alter the state of the
+// Allocator. It exists so that Allocator changes can be abandoned
+// without complex rollback logic. To commit changes to the Allocator, call
+// the proposal's Commit function
+type Proposal interface {
+	// Ports returns the port assignments generated by this proposal
+	Ports() []*api.PortConfig
+
+	// Commit commits the result of the Allocator transaction and alters
+	// the state of the port allocator. Part of the contract of the
+	// Allocator is that it should always return a valid proposal, and so
+	// Commit will always succeed and does not need to return an error.
+	Commit()
+
+	// IsNoop returns true if the Proposal doesn't actually modify the net
+	// allocator state. IsNoop does not mean the ports for an endpoint haven't
+	// changed. It only means that the publish ports marked in use by the
+	// allocator haven't changed.
+	//
+	// Used mostly for testing, and should probably not be relied on outside of
+	// this package.
+	IsNoop() bool
+}
```

Comment:
You are creating an interface in order not to export the type itself so that `proposals` can not be created directly. I would go even further and create an `Allocator` interface and extract both interfaces to the `allocator` package so that all allocators need to follow these interfaces. That would requirte some generalization.

Something in the lines of:
```
type Allocator interface {
    Allocate(old, new *api.Task) (Proposal, error)
    Deallocate(*api.Task) (Proposal, error)
}

type Proposal interface {
    Commit()
    IsNOOP() bool
}
```
And then if you want to still hide the proposal object you can make a specialized interface inside the `ports` subpackage:
```
type PortProposal interface {
    allocator.Proposal
    Ports() []*api.PortConfig
}
```